### ISSUE TITLE
Revise design of filling members

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#969]
Current implementation has problems:

- #946 (in particular https://github.com/ballerina-platform/nballerina/issues/946#issuecomment-1067601202)
- #967 
- #968 

We need a revised design to fix these.

**Comment by jclark** [comment]
Concept is to have FillerDesc object in C

```
typedef struct FillerDesc {
    TaggedPtr (*create)(struct FillerDesc *, MemberType, bool *hasIdentityPtr);
} *FillerDescPtr;
```

The fillerDesc field in ListDesc and MappingDesc changes to have type FillerDescPtr. The `hasIdentityPtr` argument gets a boolean saying whether the returned TaggedPtr has a storage identity; if this is false, then the returned value can be used to fill in multiple members.  We determine fillability of a type at compile-time: if the rest type is not fillable, then the fillerDesc field will be NULL.

The runtime library defines const FillerDesc objects for most cases:

1. nil
2. false
3. int 0
4. float 0
5. decimal 0
6. empty string
7. singleton true
8. singleton int (extracts from member type)
9. singleton decimal
10. singleton float
11. singleton string

For all of the above hasIdentityPtr will be false.

For mutable list and mapping type that is fillable and used as rest-type, init.bal will initialize a FillerDesc object.

- for mapping types, the FillerDesc will contain just a MappingDescPtr (which in turn contains a FillerDesc for the rest type)
- for array types (with no required members), the FillerDesc will similarly just contain a ListDescPtr
- for tuple types, the FillerDesc will,  in addition to a ListDescPtr, recursively contain a FillerDesc for each required member type

With tuples, the runtime uses the MemberType together with FillerDesc to recursively create the necessary nested structure.

When using constructing a value with a tuple type using a list constructor that omits required members:

- the front end checks that the omitted members are fillable
- the backend emits calls to initialize omitted members
   - for values that do not have identity, there can be an _bal_list_init_range function that initializes a range of indices to the same TaggedPtr
   - for values that do have identity, there can be a similar function that takes a FillerDescPtr; this can then reuse the same code for creating the nested value

This means we have an additional `USED_FILLER` value for `TypeHowUsed`.

**Comment by jclark** [comment]
The types module should have a function that says whether a semtype is fillable.

When it is filled with simple or string value, then this function can return that value.  For mutable values, it can return a ListAtomicType or MappingAtomicType.  For tuples, maybe also the filler for the members.

This can be cached in t:Context.

**Comment by heshanpadmasiri** [comment]
Closed with #1153
